Browsing message logs does not work with a H2 database

**Describe the bug**
I have a Frank configuration with a message log backed by table ibisstore. Both the Frank!Console and an MessageStoreListener fail to read from the queue.

**Reporter**
Martijn Dirkse

**Category**
 - Frank Framework

**To Reproduce**
Steps to reproduce the behavior:
1. Start the Frank!Runner with Frank2Example3.
2. Using Test Pipeline, issue an arbitrary message to Adapter 01.
3. Read the message log of adapter 1. You see an error in the Frank!Console about a JdbcSQLNonTransientException.
4. Look at the logfile. Adapter 03 fails to read the message from the message store.

**Expected behavior**
You do not see the error in the Frank!Console, and Adapter 03 can read the message from the message store.

**Screenshots**
I have the part of the logfile that shows the error of Adapter 03.

2020-07-14 17:19:00,299 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] [mid [1]] core.Adapter - Adapter [03 TestMessageStoreListenerAndErrorStorage] messageId [1] duration [10ms] got exit-state [success] and result [String: message 1] from PipeLine
2020-07-14 17:19:00,299 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] [mid [1]] core.Adapter - Adapter: [03 TestMessageStoreListenerAndErrorStorage] STAT: Finished processing message with messageId [1] exit-state [success] started 2020-07-14 17:19:00.288 finished 2020-07-14 17:19:00.299 total duration: 11 msecs
2020-07-14 17:19:00,299 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] null receivers.GenericReceiver - Receiver [TestMessageStoreListenerAndErrorStorage] received result: exitState [success], result [message 1]
2020-07-14 17:19:00,299 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] null receivers.GenericReceiver - Receiver [TestMessageStoreListenerAndErrorStorage] canceling TimeoutGuard, isInterrupted [false]
2020-07-14 17:19:00,299 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] null receivers.GenericReceiver - Receiver [TestMessageStoreListenerAndErrorStorage] caching first result for messageId [1]
2020-07-14 17:19:00,301 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] null jdbc.MessageStoreListener - setting parameter 1 to [1]
2020-07-14 17:19:00,303 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] null receivers.GenericReceiver - Receiver [TestMessageStoreListenerAndErrorStorage] finishes processing message
2020-07-14 17:19:00,303 DEBUG [TestMessageStoreListenerAndErrorStorage-listener[1]] null receivers.GenericReceiver - Receiver [TestMessageStoreListenerAndErrorStorage] messageId [1] correlationId [1] returning result [message 1]
2020-07-14 17:19:08,898 DEBUG [http-bio-80-exec-2] null flow.FlowDiagramManager - retrieve flow file for name[_ALL_] in folder[build\apache-tomcat-7.0.100\temp\Frank2Example3\flow\config]
2020-07-14 17:19:12,420 DEBUG [http-bio-80-exec-2] null flow.FlowDiagramManager - retrieve flow file for name[01 TestMessageLog] in folder[build\apache-tomcat-7.0.100\temp\Frank2Example3\flow\adapter]
2020-07-14 17:19:15,890 DEBUG [http-bio-80-exec-2] null jdbc.JdbcTransactionalStorage - preparing selectListQuery [SELECT MESSAGEKEY,MESSAGEID,CORRELATIONID,MESSAGEDATE,EXPIRYDATE,TYPE,HOST,LABEL,COMMENTS FROM IBISSTORE WHERE SLOTID=? AND TYPE=? ORDER BY MESSAGEDATE DESC ]
2020-07-14 17:19:15,894 ERROR [http-bio-80-exec-2] null api.ApiException - (JdbcSQLNonTransientException) SQLState [90007], errorCode [90007]: The object is already closed [90007-200]
nl.nn.adapterframework.core.ListenerException: (JdbcSQLNonTransientException) SQLState [90007], errorCode [90007]: The object is already closed [90007-200]
      at nl.nn.adapterframework.jdbc.JdbcTransactionalStorage$ResultSetIterator.advance(JdbcTransactionalStorage.java:972) ~[ibis-adapterframework-core-7.6-20200714.113155.jar:7.6-20200714.113155]
      at nl.nn.adapterframework.jdbc.JdbcTransactionalStorage$ResultSetIterator.hasNext(JdbcTransactionalStorage.java:979) ~[ibis-adapterframework-core-7.6-20200714.113155.jar:7.6-20200714.113155]
      at nl.nn.adapterframework.webcontrol.api.TransactionalStorage.getMessages(TransactionalStorage.java:581) [ibis-adapterframework-core-7.6-20200714.113155.jar:7.6-20200714.113155]
      at nl.nn.adapterframework.webcontrol.api.TransactionalStorage.browsePipeMessages(TransactionalStorage.java:445) [ibis-adapterframework-core-7.6-20200714.113155.jar:7.6-20200714.113155]
      at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_252]
      at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_252]
      at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
      at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
      at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:180) [cxf-core-3.1.18.jar:3.1.18]
      at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) [cxf-core-3.1.18.jar:3.1.18]
      at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:191) [cxf-rt-frontend-jaxrs-3.1.18.jar:3.1.18]
      at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:101) [cxf-rt-frontend-jaxrs-3.1.18.jar:3.1.18]
      at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59) [cxf-core-3.1.18.jar:3.1.18]
      at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96) [cxf-core-3.1.18.jar:3.1.18]
      at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:309) [cxf-core-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121) [cxf-core-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:189) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:303) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doGet(AbstractHTTPServlet.java:227) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at javax.servlet.http.HttpServlet.service(HttpServlet.java:621) [servlet-api.jar:?]
      at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:278) [cxf-rt-transports-http-3.1.18.jar:3.1.18]
      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303) [catalina.jar:7.0.100]
      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208) [catalina.jar:7.0.100]
      at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat7-websocket.jar:7.0.100]
      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241) [catalina.jar:7.0.100]
      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208) [catalina.jar:7.0.100]
      at nl.nn.adapterframework.webcontrol.LoginFilter.doFilter(LoginFilter.java:226) [ibis-adapterframework-core-7.6-20200714.113155.jar:7.6-20200714.113155]
      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241) [catalina.jar:7.0.100]
      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208) [catalina.jar:7.0.100]
      at nl.nn.adapterframework.http.CacheControlFilter.doFilter(CacheControlFilter.java:55) [ibis-adapterframework-core-7.6-20200714.113155.jar:7.6-20200714.113155]
      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241) [catalina.jar:7.0.100]
      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208) [catalina.jar:7.0.100]
      at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219) [catalina.jar:7.0.100]
      at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:110) [catalina.jar:7.0.100]
      at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:492) [catalina.jar:7.0.100]
      at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:165) [catalina.jar:7.0.100]
      at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:104) [catalina.jar:7.0.100]
      at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:1025) [catalina.jar:7.0.100]
      at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116) [catalina.jar:7.0.100]
      at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:452) [catalina.jar:7.0.100]
      at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1195) [tomcat-coyote.jar:7.0.100]
      at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:654) [tomcat-coyote.jar:7.0.100]
      at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:319) [tomcat-coyote.jar:7.0.100]
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
      at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-coyote.jar:7.0.100]
      at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: The object is already closed [90007-200]
      at org.h2.message.DbException.getJdbcSQLException(DbException.java:505) ~[h2-1.4.200.jar:1.4.200]
      at org.h2.message.DbException.getJdbcSQLException(DbException.java:429) ~[h2-1.4.200.jar:1.4.200]
      at org.h2.message.DbException.get(DbException.java:205) ~[h2-1.4.200.jar:1.4.200]
      at org.h2.message.DbException.get(DbException.java:181) ~[h2-1.4.200.jar:1.4.200]
      at org.h2.message.DbException.get(DbException.java:170) ~[h2-1.4.200.jar:1.4.200]
      at org.h2.jdbc.JdbcResultSet.checkClosed(JdbcResultSet.java:3223) ~[h2-1.4.200.jar:1.4.200]
      at org.h2.jdbc.JdbcResultSet.next(JdbcResultSet.java:134) ~[h2-1.4.200.jar:1.4.200]
      at nl.nn.adapterframework.jdbc.JdbcTransactionalStorage$ResultSetIterator.advance(JdbcTransactionalStorage.java:969) ~[ibis-adapterframework-core-7.6-20200714.113155.jar:7.6-20200714.113155]
      ... 50 more

**Environment:**
 - Frank!Runner
 - Microsoft Edge
 - FF! 7.6-20200714.113155

**Additional context**
Add any other context about the problem here. (f.e. ladybug report / test adapter with larva test)

--- a/core/src/main/java/nl/nn/adapterframework/core/IMessageBrowsingIterator.java
+++ b/core/src/main/java/nl/nn/adapterframework/core/IMessageBrowsingIterator.java
@@ -21,10 +21,12 @@ package nl.nn.adapterframework.core;
  * @author  Gerrit van Brakel
  * @since   4.3
  */
-public interface IMessageBrowsingIterator {
+public interface IMessageBrowsingIterator extends AutoCloseable {
 
 	boolean hasNext() throws ListenerException;
 	IMessageBrowsingIteratorItem  next() throws ListenerException;
-	void    close() throws ListenerException;
+
+	@Override
+	void  close() throws ListenerException;
 
 }

--- a/core/src/main/java/nl/nn/adapterframework/jdbc/JdbcQuerySenderBase.java
+++ b/core/src/main/java/nl/nn/adapterframework/jdbc/JdbcQuerySenderBase.java
@@ -294,8 +294,10 @@ public abstract class JdbcQuerySenderBase<H> extends JdbcSenderBase<H> {
 			} finally {
 				closeStatementSet(queryExecutionContext, session);
 			}
+		} catch (SenderException|TimeOutException e) {
+			throw e;
 		} catch (Exception e) {
-			throw new SenderException(getLogPrefix() + "cannot return connection to repeatable read", e);
+			throw new SenderException(e);
 		}
 	}
 	

--- a/core/src/main/java/nl/nn/adapterframework/jdbc/JdbcTableListener.java
+++ b/core/src/main/java/nl/nn/adapterframework/jdbc/JdbcTableListener.java
@@ -36,6 +36,7 @@ public class JdbcTableListener extends JdbcListener {
 	private String statusValueProcessed;
 	private String statusValueError;
 	
+	@Override
 	public void configure() throws ConfigurationException {
 		if (StringUtils.isEmpty(getTableName())) {
 			throw new ConfigurationException(getLogPrefix()+"must specifiy tableName");
@@ -76,6 +77,7 @@ public class JdbcTableListener extends JdbcListener {
 				" WHERE "+getKeyField()+"=?";
 	}
 
+	@Override
 	public String getPhysicalDestinationName() {
 		return super.getPhysicalDestinationName()+" "+getTableName();
 	}
@@ -91,11 +93,13 @@ public class JdbcTableListener extends JdbcListener {
 	}
 
 	@IbisDoc({"primary key field of the table, used to identify messages", ""})
+	@Override
 	public void setKeyField(String fieldname) {
 		super.setKeyField(fieldname);
 	}
 
 	@IbisDoc({"(optional) field containing the message data", "<i>same as keyfield</i>"})
+	@Override
 	public void setMessageField(String fieldname) {
 		super.setMessageField(fieldname);
 	}

--- a/core/src/main/java/nl/nn/adapterframework/jdbc/JdbcTransactionalStorage.java
+++ b/core/src/main/java/nl/nn/adapterframework/jdbc/JdbcTransactionalStorage.java
@@ -550,8 +550,7 @@ public class JdbcTransactionalStorage<S extends Serializable> extends JdbcFacade
 	 *	Checks if table exists, and creates when necessary. 
 	 */
 	public void initialize(IDbmsSupport dbmsSupport) throws JdbcException, SQLException, SenderException {
-		Connection conn = getConnection();
-		try {
+		try (Connection conn = getConnection()) {
 			boolean tableMustBeCreated;
 
 			if (checkIfTableExists) {
@@ -572,16 +571,11 @@ public class JdbcTransactionalStorage<S extends Serializable> extends JdbcFacade
 
 			if (isCreateTable() && tableMustBeCreated || forceCreateTable) {
 				log.info(getLogPrefix()+"creating table ["+getPrefix()+getTableName()+"] for transactional storage");
-				Statement stmt = conn.createStatement();
-				try {
+				try (Statement stmt = conn.createStatement()) {
 					createStorage(conn, stmt, dbmsSupport);
-				} finally {
-					stmt.close();
-					conn.commit();
-				}
+				} 
+				conn.commit();
 			}
-		} finally {
-			conn.close();
 		}
 	}
 
@@ -1010,28 +1004,42 @@ public class JdbcTransactionalStorage<S extends Serializable> extends JdbcFacade
 
 	@Override
 	public IMessageBrowsingIterator getIterator(Date startTime, Date endTime, SortOrder order) throws ListenerException {
-		try (Connection conn = getConnection()) {
+		Connection conn;
+		PreparedStatement stmt=null;
+		IMessageBrowsingIterator result=null;
+		try {
+			conn = getConnection();
+		} catch (JdbcException e) {
+			throw new ListenerException(e);
+		}
+		try {
 			String query = getSelectListQuery(getDbmsSupport(), startTime, endTime, order);
 			if (log.isDebugEnabled()) {
 				log.debug("preparing selectListQuery ["+query+"]");
 			}
-			try (PreparedStatement stmt = conn.prepareStatement(query)) {
-				if (startTime==null && endTime==null) {
-					applyStandardParameters(stmt, false, false);
-				} else {
-					int paramPos=applyStandardParameters(stmt, true, false);
-					if (startTime!=null) {
-						stmt.setTimestamp(paramPos++, new Timestamp(startTime.getTime()));
-					}
-					if (endTime!=null) {
-						stmt.setTimestamp(paramPos++, new Timestamp(endTime.getTime()));
-					}
+			stmt = conn.prepareStatement(query);
+			if (startTime==null && endTime==null) {
+				applyStandardParameters(stmt, false, false);
+			} else {
+				int paramPos=applyStandardParameters(stmt, true, false);
+				if (startTime!=null) {
+					stmt.setTimestamp(paramPos++, new Timestamp(startTime.getTime()));
 				}
-				ResultSet rs =  stmt.executeQuery();
-				return new ResultSetIterator(conn,rs);
+				if (endTime!=null) {
+					stmt.setTimestamp(paramPos++, new Timestamp(endTime.getTime()));
+				}
 			}
-		} catch (Exception e) {
+			ResultSet rs =  stmt.executeQuery();
+			result = new ResultSetIterator(conn,rs);
+			return result;
+		} catch (SQLException e) {
 			throw new ListenerException(e);
+		} finally {
+			if (result==null) {
+				// in happy flow IMessageBrowsingIterator will close resources.
+				// If it has not been created, we'll have to close them here.
+				JdbcUtil.fullClose(conn, stmt);
+			}
 		}
 	}
 

--- a/core/src/main/java/nl/nn/adapterframework/webcontrol/api/BrowseJdbcTable.java
+++ b/core/src/main/java/nl/nn/adapterframework/webcontrol/api/BrowseJdbcTable.java
@@ -70,13 +70,13 @@ public final class BrowseJdbcTable extends Base {
 				datasource = entry.getValue().toString();
 			}
 			if(key.equalsIgnoreCase("table")) {
-				tableName = entry.getValue().toString().toLowerCase();
+				tableName = entry.getValue().toString();
 			}
 			if(key.equalsIgnoreCase("where")) {
-				where = entry.getValue().toString().toLowerCase();
+				where = entry.getValue().toString();
 			}
 			if(key.equalsIgnoreCase("order")) {
-				order = entry.getValue().toString().toLowerCase();
+				order = entry.getValue().toString();
 			}
 			if(key.equalsIgnoreCase("rowNumbersOnly")) {
 				rowNumbersOnly = Boolean.parseBoolean(entry.getValue().toString());
@@ -123,71 +123,68 @@ public final class BrowseJdbcTable extends Base {
 			qs.setName("QuerySender");
 			qs.setDatasourceName(datasource);
 
-			//if (form_numberOfRowsOnly || qs.getDatabaseType() == DbmsSupportFactory.DBMS_ORACLE) {
-				qs.setQueryType("select");
-				qs.setBlobSmartGet(true);
-				qs.setIncludeFieldDefinition(true);
-				qs.configure(true);
-				qs.open();
+			qs.setQueryType("select");
+			qs.setSqlDialect("Oracle");
+			qs.setBlobSmartGet(true);
+			qs.setIncludeFieldDefinition(true);
+			qs.configure(true);
+			qs.open();
 
-				try (Connection conn =qs.getConnection()) {
-					ResultSet rs = null;
-					try {
-						rs = conn.getMetaData().getColumns(null, null, tableName, null);
-						if (!rs.isBeforeFirst()) {
-							rs.close();
-							rs = conn.getMetaData().getColumns(null, null, tableName.toUpperCase(), null);
-						}
-		
-						String fielddefinition = "<fielddefinition>";
-						while(rs.next()) {
-							String field = "<field name=\"" + rs.getString(4) + "\" type=\"" + DB2XMLWriter.getFieldType(rs.getInt(5)) + "\" size=\"" + rs.getInt(7) + "\"/>";
-							fielddefinition = fielddefinition + field;
-							fieldDef.put(rs.getString(4), DB2XMLWriter.getFieldType(rs.getInt(5)) + "("+rs.getInt(7)+")");
-						}
-						fielddefinition = fielddefinition + "</fielddefinition>";
-		
-						String browseJdbcTableExecuteREQ =
-							"<browseJdbcTableExecuteREQ>"
-								+ "<dbmsName>"
-								+ qs.getDbmsSupport().getDbmsName()
-								+ "</dbmsName>"
-								+ "<tableName>"
-								+ tableName
-								+ "</tableName>"
-								+ "<where>"
-								+ XmlUtils.encodeChars(where)
-								+ "</where>"
-								+ "<numberOfRowsOnly>"
-								+ rowNumbersOnly
-								+ "</numberOfRowsOnly>"
-								+ "<order>"
-								+ order
-								+ "</order>"
-								+ "<rownumMin>"
-								+ minRow
-								+ "</rownumMin>"
-								+ "<rownumMax>"
-								+ maxRow
-								+ "</rownumMax>"
-								+ fielddefinition
-								+ "<maxColumnSize>1000</maxColumnSize>"
-								+ "</browseJdbcTableExecuteREQ>";
-						URL url = ClassUtils.getResourceURL(getClassLoader(), DB2XML_XSLT);
-						if (url != null) {
-							Transformer t = XmlUtils.createTransformer(url);
-							query = XmlUtils.transformXml(t, browseJdbcTableExecuteREQ);
-						}
-						result = qs.sendMessage(new Message(query), null).asString();
-					} finally {
-						if (rs!=null) {
-							rs.close();
-						}
+			try (Connection conn =qs.getConnection()) {
+				ResultSet rs = null;
+				try {
+					rs = conn.getMetaData().getColumns(null, null, tableName, null);
+					if (!rs.isBeforeFirst()) {
+						rs.close();
+						rs = conn.getMetaData().getColumns(null, null, tableName.toUpperCase(), null);
+					}
+	
+					String fielddefinition = "<fielddefinition>";
+					while(rs.next()) {
+						String field = "<field name=\"" + rs.getString(4) + "\" type=\"" + DB2XMLWriter.getFieldType(rs.getInt(5)) + "\" size=\"" + rs.getInt(7) + "\"/>";
+						fielddefinition = fielddefinition + field;
+						fieldDef.put(rs.getString(4), DB2XMLWriter.getFieldType(rs.getInt(5)) + "("+rs.getInt(7)+")");
+					}
+					fielddefinition = fielddefinition + "</fielddefinition>";
+	
+					String browseJdbcTableExecuteREQ =
+						"<browseJdbcTableExecuteREQ>"
+							+ "<dbmsName>"
+							+ qs.getDbmsSupport().getDbmsName()
+							+ "</dbmsName>"
+							+ "<tableName>"
+							+ tableName
+							+ "</tableName>"
+							+ "<where>"
+							+ XmlUtils.encodeChars(where)
+							+ "</where>"
+							+ "<numberOfRowsOnly>"
+							+ rowNumbersOnly
+							+ "</numberOfRowsOnly>"
+							+ "<order>"
+							+ order
+							+ "</order>"
+							+ "<rownumMin>"
+							+ minRow
+							+ "</rownumMin>"
+							+ "<rownumMax>"
+							+ maxRow
+							+ "</rownumMax>"
+							+ fielddefinition
+							+ "<maxColumnSize>1000</maxColumnSize>"
+							+ "</browseJdbcTableExecuteREQ>";
+					URL url = ClassUtils.getResourceURL(getClassLoader(), DB2XML_XSLT);
+					if (url != null) {
+						Transformer t = XmlUtils.createTransformer(url);
+						query = XmlUtils.transformXml(t, browseJdbcTableExecuteREQ);
+					}
+					result = qs.sendMessage(new Message(query), null).asString();
+				} finally {
+					if (rs!=null) {
+						rs.close();
 					}
 				}
-			//} else {
-				//error("errors.generic","This function only supports oracle databases",null);
-			//}
+			}
 		} catch (Throwable t) {
 			throw new ApiException("An error occured on executing jdbc query ["+query+"]", t);
 		} finally {

--- a/core/src/main/java/nl/nn/adapterframework/webcontrol/api/BrowseQueue.java
+++ b/core/src/main/java/nl/nn/adapterframework/webcontrol/api/BrowseQueue.java
@@ -93,8 +93,6 @@ public final class BrowseQueue extends Base {
 		if(type == null)
 			throw new ApiException("No type provided");
 
-		IMessageBrowsingIterator it = null;
-
 		try {
 			JmsBrowser<javax.jms.Message> jmsBrowser = new JmsBrowser<>();
 			jmsBrowser.setName("BrowseQueueAction");
@@ -102,29 +100,30 @@ public final class BrowseQueue extends Base {
 			jmsBrowser.setDestinationName(destination);
 			jmsBrowser.setDestinationType(type);
 
-			it = jmsBrowser.getIterator();
 			List<Map<String, Object>> messages = new ArrayList<Map<String, Object>>();
-			while (it.hasNext()) {
-				IMessageBrowsingIteratorItem item = it.next();
-				Map<String, Object> message = new HashMap<String, Object>();
-				message.put("comment", item.getCommentString());
-				message.put("correlationId", item.getCorrelationId());
-				try {
-					message.put("expiryDate", item.getExpiryDate());
-				} catch (Exception e) {
-					log.warn("Could not get expiryDate",e);
+			try (IMessageBrowsingIterator it = jmsBrowser.getIterator()) {
+				while (it.hasNext()) {
+					IMessageBrowsingIteratorItem item = it.next();
+					Map<String, Object> message = new HashMap<String, Object>();
+					message.put("comment", item.getCommentString());
+					message.put("correlationId", item.getCorrelationId());
+					try {
+						message.put("expiryDate", item.getExpiryDate());
+					} catch (Exception e) {
+						log.warn("Could not get expiryDate",e);
+					}
+					message.put("host", item.getHost());
+					message.put("id", item.getId());
+					try {
+						message.put("insertDate", item.getInsertDate());
+					} catch (Exception e) {
+						log.warn("Could not get insertDate",e);
+					}
+					message.put("type", item.getType());
+					message.put("label", item.getLabel());
+	
+					messages.add(message);
 				}
-				message.put("host", item.getHost());
-				message.put("id", item.getId());
-				try {
-					message.put("insertDate", item.getInsertDate());
-				} catch (Exception e) {
-					log.warn("Could not get insertDate",e);
-				}
-				message.put("type", item.getType());
-				message.put("label", item.getLabel());
-
-				messages.add(message);
 			}
 
 			log.debug("Browser returned " + messages.size() + " messages");
@@ -134,14 +133,6 @@ public final class BrowseQueue extends Base {
 		}
 		catch (Exception e) {
 			throw new ApiException("Error occured browsing messages", e);
-		}
-		finally {
-			try {
-				if (it!=null)
-					it.close();
-			} catch (ListenerException e) {
-				log.error(e);
-			}
 		}
 
 		return Response.status(Response.Status.OK).entity(returnMap).build();

--- a/core/src/main/java/nl/nn/adapterframework/webcontrol/api/BrowseQueue.java
+++ b/core/src/main/java/nl/nn/adapterframework/webcontrol/api/BrowseQueue.java
@@ -34,7 +34,6 @@ import javax.ws.rs.core.Response;
 
 import nl.nn.adapterframework.core.IMessageBrowsingIterator;
 import nl.nn.adapterframework.core.IMessageBrowsingIteratorItem;
-import nl.nn.adapterframework.core.ListenerException;
 import nl.nn.adapterframework.jms.JmsBrowser;
 import nl.nn.adapterframework.jms.JmsRealmFactory;
 

--- a/core/src/main/java/nl/nn/adapterframework/webcontrol/api/ExecuteJdbcQuery.java
+++ b/core/src/main/java/nl/nn/adapterframework/webcontrol/api/ExecuteJdbcQuery.java
@@ -95,6 +95,7 @@ public final class ExecuteJdbcQuery extends Base {
 	public Response execute(LinkedHashMap<String, Object> json) throws ApiException {
 
 		String datasource = null, resultType = null, query = null, queryType = null, result = "", returnType = MediaType.APPLICATION_XML;
+		boolean dirtyRead = false;
 		for (Entry<String, Object> entry : json.entrySet()) {
 			String key = entry.getKey();
 			if(key.equalsIgnoreCase("datasource")) {
@@ -108,6 +109,9 @@ public final class ExecuteJdbcQuery extends Base {
 				if(resultType.equalsIgnoreCase("json")) {
 					returnType = MediaType.APPLICATION_JSON;
 				}
+			}
+			if(key.equalsIgnoreCase("dirtyRead")) {
+				dirtyRead = Boolean.parseBoolean(entry.getValue().toString());
 			}
 			if(key.equalsIgnoreCase("query")) {
 				query = entry.getValue().toString();
@@ -145,6 +149,7 @@ public final class ExecuteJdbcQuery extends Base {
 			qs.setName("QuerySender");
 			qs.setDatasourceName(datasource);
 			qs.setQueryType(queryType);
+			qs.setDirtyRead(dirtyRead);
 			qs.setBlobSmartGet(true);
 			qs.configure(true);
 			qs.open();

--- a/core/src/main/java/nl/nn/adapterframework/webcontrol/api/TransactionalStorage.java
+++ b/core/src/main/java/nl/nn/adapterframework/webcontrol/api/TransactionalStorage.java
@@ -570,53 +570,43 @@ public class TransactionalStorage extends Base {
 
 		Date startDate = null;
 		Date endDate = null;
-		try {
-			IMessageBrowsingIterator iterator = null;
-			try {
-				iterator = transactionalStorage.getIterator(startDate, endDate, filter.getSortOrder());
-	
-				int count;
-				List<Object> messages = new LinkedList<Object>();
-				
-				for (count=0; iterator.hasNext(); ) {
-					IMessageBrowsingIteratorItem iterItem = iterator.next();
-					try {
-						if(!filter.matchAny(iterItem))
-							continue;
+		try (IMessageBrowsingIterator iterator = transactionalStorage.getIterator(startDate, endDate, filter.getSortOrder())) {
+			int count;
+			List<Object> messages = new LinkedList<Object>();
+			
+			for (count=0; iterator.hasNext(); ) {
+				IMessageBrowsingIteratorItem iterItem = iterator.next();
+				try {
+					if(!filter.matchAny(iterItem))
+						continue;
 
-						count++;
-						if (count > filter.skipMessages()) { 
-							Map<String, Object> message = new HashMap<String, Object>(3);
+					count++;
+					if (count > filter.skipMessages()) { 
+						Map<String, Object> message = new HashMap<String, Object>(3);
 
-							message.put("id", iterItem.getId());
-							message.put("pos", count);
-							message.put("originalId", iterItem.getOriginalId());
-							message.put("correlationId", iterItem.getCorrelationId());
-							message.put("type", iterItem.getType());
-							message.put("host", iterItem.getHost());
-							message.put("insertDate", iterItem.getInsertDate());
-							message.put("expiryDate", iterItem.getExpiryDate());
-							message.put("comment", iterItem.getCommentString());
-							message.put("label", iterItem.getLabel());
-							messages.add(message);
-						}
-		
-						if (filter.maxMessages() > 0 && count >= (filter.maxMessages() + filter.skipMessages())) {
-							log.warn("stopped iterating messages after ["+count+"]: limit reached");
-							break;
-						}
-					} finally {
-						iterItem.release();
+						message.put("id", iterItem.getId());
+						message.put("pos", count);
+						message.put("originalId", iterItem.getOriginalId());
+						message.put("correlationId", iterItem.getCorrelationId());
+						message.put("type", iterItem.getType());
+						message.put("host", iterItem.getHost());
+						message.put("insertDate", iterItem.getInsertDate());
+						message.put("expiryDate", iterItem.getExpiryDate());
+						message.put("comment", iterItem.getCommentString());
+						message.put("label", iterItem.getLabel());
+						messages.add(message);
 					}
+	
+					if (filter.maxMessages() > 0 && count >= (filter.maxMessages() + filter.skipMessages())) {
+						log.warn("stopped iterating messages after ["+count+"]: limit reached");
+						break;
+					}
+				} finally {
+					iterItem.release();
 				}
-				returnObj.put("messages", messages);
 			}
-			finally {
-				if(iterator != null)
-					iterator.close();
-			}
-		}
-		catch (ListenerException|IOException e) {
+			returnObj.put("messages", messages);
+		} catch (ListenerException|IOException e) {
 			throw new ApiException(e);
 		}
 

--- a/webapp/src/main/webapp/iaf/gui/views/ExecuteJdbcQuery.html
+++ b/webapp/src/main/webapp/iaf/gui/views/ExecuteJdbcQuery.html
@@ -30,6 +30,12 @@
                             </div>
                         </div>
                         <div class="form-group">
+                            <label class="col-sm-3 control-label">Dirty read</label>
+                            <div class="col-sm-9">
+                                <span class="form-control m-b" style="border: none;"><input icheck type="checkbox" ng-model="form.dirtyRead" /></span>
+                            </div>
+                        </div>
+                        <div class="form-group">
                             <label class="col-sm-3 control-label">Query</label>
                             <div class="col-sm-9">
                                 <textarea class="form-control" quick-submit-form="submit(form)" name="query" rows="5" style="resize:none;" ng-model="form.query"></textarea>

